spec.bs

```diff
@@ -2921,6 +2926,188 @@ This specification defines two [=policy-controlled features=] identified by the
+
+NOTE: This is only captured during a {{WindowOrWorkerGlobalScope/fetch()}} call with the
+`{adAuctionHeaders: true}` option.
+
+<dl dfn-for="direct from seller signals key">
+: <dfn>seller</dfn>
+:: An [=origin=]. Matches the origin that served the captured [:Ad-Auction-Signals:] header.
+: <dfn>ad slot</dfn>
+:: A [=string=]. Matches the `adSlot` key of the JSON dictionaries in the top-level array of the
+  [:Ad-Auction-Signals:] value.
+
+</dl>
+
+<h3 id=direct-from-sellers-signals-struct>Direct from seller signals</h3>
+A <dfn>direct from seller signals</dfn> is a [=struct=] with the following [=struct/items=]:
+
+NOTE: This is only captured during a {{WindowOrWorkerGlobalScope/fetch()}} call with the
+`{adAuctionHeaders: true}` option.
+
+<dl dfn-for="direct from seller signals">
+: <dfn>seller signals</dfn>
+:: Null or a [=string=].
+  Opaque JSON data passed to both buyers' and the seller's [=script runners=].
+: <dfn>auction signals</dfn>
+:: Null or a [=string=].
+  Opaque JSON data passed to the seller's [=script runner=].
+: <dfn>per buyer signals</dfn>
+:: A [=map=] whose [=map/keys=] are [=origins=] and whose [=map/values=] are [=strings=].
+  [=map/Keys=] are buyers and must be valid HTTPS origins. [=map/Values=] are opaque JSON data
+  passed to corresponding buyer's [=script runner=].
+
+</dl>
+
+Each [=traversable navigable=] has a <dfn>captured ad auction headers</dfn>, which is a [=map=]
+whose [=map/keys=] are [=direct from seller signals keys=] and whose [=map/values=] are
+[=direct from seller signals=].
+</div>
+
+<div algorithm="fetch auction headers patch">
+The following will be added to [[Fetch#http-extensions]]:
+
+<h3 id=sec-ad-auction-fetch-header>`Sec-Ad-Auction-Fetch` header</h3>
+
+The <dfn http-header><code>Sec-Ad-Auction-Fetch</code></dfn> request header is an optional
+[=structured header=] with of type [=structured header/boolean=]. If equal to `?1`, the user agent
+will remove any [:Ad-Auction-Signals:] from the response returned to
+{{WindowOrWorkerGlobalScope/fetch()}} -- the [:Ad-Auction-Signals:] value will instead only be used
+in Protected Audiences auctions. [:Sec-Ad-Auction-Fetch:] will only be set if the
+{{WindowOrWorkerGlobalScope/fetch()}} call was made with the `{adAuctionHeaders: true}` option.
+
+<h3 id=ad-auction-signals-header>`Ad-Auction-Signals` header</h3>
+
+The <dfn http-header><code>Ad-Auction-Signals</code></dfn> response header provides value
+of a JSON array of dictionaries, each with an `adSlot` key. Protected Audience non-component,
+component, and top-level auctions may specify which signals to load by the `adSlot` key.
+</div>
+
+<div algorithm="fetch auction signals response patch">
+The following step will be added to the [=http-network-or-cache fetch=] algorithm, before step
+"Return |response|.":
+
+1. If |response| is not null, then run [=update captured headers=] with [=this=]'s

```

Comment:
Unfortunately this step is pretty broken. This algorithm doesn't have a `[=this=]` available in it, since the algorithm itself is not a platform object. That means we can't reach into it ("this") and get the global, document, and so on. Instead, you have two options that I can see:

1. Do this on the "main thread", by modifying the steps in https://fetch.spec.whatwg.org/#ref-for-process-response%E2%91%A0 which run on the document-that-initiated-the-fetch-and-will-receive-its-response's thread. This is definitely the most straigthforward, but if you have a security requirement of not letting the header reach the "renderer process", then this might not be what you want. Do you have that requirement? I see you mention something along these lines in the first "Note" in https://pr-preview.s3.amazonaws.com/WICG/turtledove/771/a5f6630...caraitto:a95a01a.html#update-captured-headers. But there's a difference between not letting normal script read a value, and not exposing it to the renderer process altogether (where spectre attacks can leave the value readable by a compromised page anyways).
2. In the same place you're doing this right now, just get the global object from https://fetch.spec.whatwg.org/#fetch-params-task-destination. So something like `If _fetchParams_'s [=fetch params/task destination=] is a [=global object=]` (you might need to mess with the syntax a little), then you can use that global object to get the document and so on. This is a little strange since we're not on the thread that the global object is tied to, yet we're reaching into it and grabbing data structures. But I think that's probably fine as long as there isn't any obvious races between when this data structure is written to and when it is read from? Unfortunately I can't quite tell where https://pr-preview.s3.amazonaws.com/WICG/turtledove/771/a5f6630...caraitto:a95a01a.html#captured-ad-auction-headers is read from. It kind of looks like it's read from somewhere in https://github.com/WICG/turtledove/pull/774 but I haven't reviewed it all yet, so maybe you could give me a hint as to whether there is a race condition between when we write to this map and when it could be read from, or if they always happen sequentially? Understanding how we handle this in the implementation would be good to know too.

Reply:
>_"the caller to fetch() needs to await the promise returned from fetch() to ensure that the signals will be available in a subsequent navigator.runAdAuction() call"_

Sorry, where is the caller that awaits the promise returned from fetch()?

Reply:
I'm not sure I understand? The ad tech code makes the call to fetch(), so it is responsible for making sure to await that promise before resolving the promise for directFromSellerSignalsHeaderAdSlot.

Reply:
Also, in the implementation, I believe the structure holding responses is protected by only being used on one sequence.

Reply:
oh ok, the caller being the ad tech's script. Misunderstood that.

Reply:
> we don't give the signals to the renderer process, although spectre-style attacks are mostly out of the threat model for this feature.

Is this true? Do we do something particularly special with the response? Typically if the server adheres to CORS, the response is simply shared with the `fetch()`-er but it sounds like you're saying the Chromium implementation does something special to strip the response? I'd be surprised by this but want to double check.

Reply:
Ugh, sorry! I'm now remembering that the headers are manually stripped from the response, which is how they don't get shared with the body. Now I'm curious though: why do we do that? If the server has opted into CORS (i.e., opted into sharing the response with the page), why not let the headers all the way through? That's basically what CORS is for.

Reply:
The issue is that the server *does* need to grant permission for the headers to be readable for the purpose of delivering them to auction worklet functions (this is what the CORS Access-Control-Allow-Origin header is doing) -- it's just that it doesn't want those signals to be available on the response accessible to script runners in the document. The purpose of this feature (directFromSellerSignals) is to provide such a mechanism.

Reply:
> it's just that it doesn't want those signals to be available on the response accessible to script runners in the document.

I'm not sure I follow this. Do you mean the server doesn't want the response to be available to _normal script_ ~runners~ in the document (unrelated to "script runner" / auction worklets)?

Reply:
Correct -- normal scripts in the document shouldn't be able to access the header.